src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,212 @@
+            // If the type containing the method is generic, synthesize a helper type and put the payload field there.
+            // If the payload field is part of a generic type, there will be a new instance of the field per instantiation of the generic,
+            // and so the payload field must be a member of another type.
+            NamedTypeSymbol containingType = method.ContainingType;
+
+            SynthesizedFieldSymbol payloadField = new SynthesizedFieldSymbol(containingType, payloadType, GeneratedNames.MakeSynthesizedInstrumentationPayloadFieldName(method, methodOrdinal), isStatic: true);
+            factory.AddField(containingType, payloadField);
+            return payloadField;
+        }
+        
+        private static MethodSymbol GetInstrumentationHelper(CSharpCompilation compilation, WellKnownMember member, DiagnosticBag diagnostics)
+        {
+            MethodSymbol helper = compilation.GetWellKnownTypeMember(member) as MethodSymbol;
+
+            if ((object)helper == null)
+            {
+                RuntimeMembers.MemberDescriptor memberDescriptor = WellKnownMembers.GetDescriptor(member);
+                diagnostics.Add(Diagnostic.Create(new CSDiagnosticInfo(ErrorCode.ERR_MissingPredefinedMember, memberDescriptor.DeclaringTypeMetadataName, memberDescriptor.Name)));
+            }
+
+            return helper;
+        }
+
+        private static MethodSymbol GetCreatePayload(CSharpCompilation compilation, DiagnosticBag diagnostics)
+        {
+            return GetInstrumentationHelper(compilation, WellKnownMember.Microsoft_CodeAnalysis_Runtime_Instrumentation__CreatePayload, diagnostics);
+        }
+
+        private static MethodSymbol GetFlushPayload(CSharpCompilation compilation, DiagnosticBag diagnostics)
+        {
+            return GetInstrumentationHelper(compilation, WellKnownMember.Microsoft_CodeAnalysis_Runtime_Instrumentation__FlushPayload, diagnostics);
+        }
+
+        private static bool IsTestMethod(MethodSymbol method)
+        {
+            // ToDo PROTOTYPE (https://github.com/dotnet/roslyn/issues/9811): Make this real. 
+            return method.Name.StartsWith("Test");
+        }
+    }
+
+    internal sealed class InstrumentationInjectionRewriter : BoundTreeRewriterWithStackGuard
+    {
+        private readonly MethodSymbol _method;
+        private readonly ArrayBuilder<SourceSpan> _spansBuilder;
+        private readonly FieldSymbol _payload;
+        private readonly TypeCompilationState _compilationState;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly DebugDocumentProvider _debugDocumentProvider;
+
+        public static void InstrumentMethod(MethodSymbol method, BoundBlock methodBody, FieldSymbol payloadField, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans, out BoundBlock newMethodBody)
```

Comment:
There are two returned results of equal importance. In such circumstances, do we generally prefer to return one and have an out parameter for the other?